intermediate_source/rpc_param_server_tutorial.rst

```diff
@@ -0,0 +1,398 @@
+
+.. code-block:: python
+
+   def get_accuracy(test_loader, model):
+       model.eval()
+       correct_sum = 0
+       # Use GPU to evaluate if possible
+       device = torch.device("cuda:0" if model.num_gpus > 0 
+           and torch.cuda.is_available() else "cpu")
+       with torch.no_grad():
+           for i, (data, target) in enumerate(test_loader):
+               out = model(data, -1)
+               pred = out.argmax(dim=1, keepdim=True)
+               pred, target = pred.to(device), target.to(device)
+               correct = pred.eq(target.view_as(pred)).sum().item()
+               correct_sum += correct
+
+       print(f"Accuracy {correct_sum / len(test_loader.dataset)}")
+
+Next, similar to how we defined ``run_parameter_server`` as the main loop for our ``ParameterServer`` that is responsible for initializing RPC, let's define a similar loop for our trainers. The difference will be that our trainers must run the training loop we defined above:
+
+.. code-block:: python
+
+   # Main loop for trainers.
+   def run_worker(rank, world_size, num_gpus, train_loader, test_loader):
+       print(f"Worker rank {rank} initializing RPC")
+       rpc.init_rpc(
+           name=f"trainer_{rank}",
+           rank=rank,
+           world_size=world_size)
+
+       print(f"Worker {rank} done initializing RPC")
+
+       run_training_loop(rank, num_gpus, train_loader, test_loader)
+       rpc.shutdown()
+
+Note that similar to ``run_parameter_server``\ , ``rpc.shutdown()`` will by default wait for all workers, both trainers and ParameterServers, to call into ``rpc.shutdown()`` before this node exits. This ensures that nodes are terminated gracefully and no node goes offline while another is expecting it to be online.
+
+We've now completed our trainer and parameter server specific code, and all that's left is to add code to launch trainers and parameter servers. First, we must take in various arguments that apply to our parameter server and trainers. ``world_size`` corresponds to the total number of nodes that will participate in training, and is the sum of all trainers and the parameter server. We also must pass in a unique ``rank`` for each individual process, from 0 (where we will run our single parameter server) to ``world_size - 1``. ``master_addr`` and ``master_port`` are arguments that can be used to identify where the rank 0 process is running, and will be used by individual nodes to discover each other. To test this example out locally, simply pass in ``localhost`` and the same ``master_port`` to all instances spawned. Note that for demonstration purposes, this example supports only between 0-2 GPUs, although the pattern can be extended to make use of additional GPUs.
+
+.. code-block:: python
+
+   if __name__ == '__main__':
+       parser = argparse.ArgumentParser(
+           description="Parameter-Server RPC based training")
+       parser.add_argument(
+           "world_size",
```

Comment:
inconsistency: some arg has -- prefix, some don't?

Reply:
With argparse prepending `--` makes the argument optional which should be the case for master_addr and master_port, for the other args, they are required.

Reply:
Will change to make all the args explicit